### ISSUE TITLE
Ability to Set Both RetryPolicy? Policy and AsyncRetryHandler? Handler in TaskRetryOptions

### ISSUE THREAD

**Issue by tomseida** [microsoft/durabletask-dotnet#279]
I am confused why `HandleAsync` was made obsolete in `RetryPolicy` via the [pull request](https://github.com/microsoft/durabletask-dotnet/commit/78300ab0e7dcd0e92847fa74d7e3f0278521c512) tied to issue #275 without providing a means to have both a  retry policy and retry handler.

Per the current implementation `TaskRetryOptions` one can have either `RetryPolicy? Policy` or `AsyncRetryHandler? Handler`, but not both. If one chooses the former, they can create sleep periods between retries, but will retry all exceptions generated by the activity function. If one chooses the later, they can control which exceptions are retried, but not the sleep period between retries.

The orchestrations I implemented prior to moving to isolated functions often used the retry policy and custom handler in tandem. Below is a simplified Activity function that throws transient and non-transient exceptions. 

```
public class AF_Example(IWidgetRepository widgetRepository, IMyApiClientService myApiClientService)
{
    [Function(nameof(AF_Example))]
    public async Task<string> HandleAsync(
        [ActivityTrigger] Guid widgetUID,
        FunctionContext executionContext,
        CancellationToken cancellationToken)
    {
        // If widget is not found, this activity function will never succeed.
        Widget widget = await widgetRepository.GetAsync(widgetUID, cancellationToken)
            ?? throw new NotFoundException(nameof(Widget), widgetUID.ToString());

        // Register widget via an API service and obtain confirmation ID. This
        // operation may experience transient exceptions (5XX status code).
        string confirmationID = await myApiClientService.RegisterWidgetAsync(widget, cancellationToken);

        return confirmationID;
    }
}
```

The first action in this activity function attempts to get a widget from a database. If the `Widget` is not found, there is no point in the orchestration retrying. I would set `AsyncRetryHandler? Handler` delegate to not retry on `NotFoundException`. My orchestrations would have a `try/catch` block to capture these exceptions for handling and I had confidence that no subsequent activity functions would be executed because of the exception.

The second action is a call to an API in which transient issue are possible. Even if  `IMyApiClientService` has retry logic (e.g., Polly), I found it useful to have my activity function implement retry policy that spans a much longer period of time than my API client service retry period. I want to set a `RetryPolicy? Policy` to create my sleep periods and constraints on how often and\or long I want to retry, and set the `AsyncRetryHandler? Handler` delegate to retry on `HttpRequestException` when its status code is 5XX. 

I think it would be beneficial to be able to set `RetryPolicy? Policy` and `AsyncRetryHandler? Handler` in `TaskRetryOptions` so `TaskOrchestrationContextWrapper.CallActivityAsync()` could schedule an activity function retry if the handler and policy specify a retry should be performed.

Hoping you will consider this feature in that it appears to me to not alter existing operations and provides a useful feature that leverages configuration classes as is (sans adding a constructor to `TaskRetryOptions`).

**Comment by davidmrdavid** [comment]
@cgillum: for comment, since he's the PR's author

**Comment by Kobudzik** [comment]
Hello!
This problem bothers me too, unfortunately.

Have anyone tried using **AsyncRetryHandler** (to filter exceptions on which to return) and using Task Delay, to make handler retry with custom periods? That was my idea how to handle retries, then I have found this **HandleAsync** property, then it turned it is not supported.

**Comment by tomseida** [comment]
@Kobudzik: I have created pull request #281 to address this valuable missing feature and I am awaiting response\action from the team at Microsoft.

**Comment by cgillum** [comment]
Responding to the opening statement of the original post:

> I am confused why HandleAsync was made obsolete in RetryPolicy via the [pull request](https://github.com/microsoft/durabletask-dotnet/commit/78300ab0e7dcd0e92847fa74d7e3f0278521c512) tied to issue https://github.com/microsoft/durabletask-dotnet/pull/275 without providing a means to have both a retry policy and retry handler.

Quoting from the PR you linked to:

> It has no implementation at all and some users are trying to use it and opening issues because it's not doing anything.

In other words, `HandleAsync` wasn't doing anything. The property was accidentally committed to the API surface as part of various refactoring when the original implementation was being written. We made it `[Obsolete]` rather than deleting it to avoid a breaking API change. It was *never* possible to have both a static retry policy and a dynamic retry policy at the same time for a single activity call.

@tomseida you mentioned the following:

> The orchestrations I implemented prior to moving to isolated functions often used the retry policy and custom handler in tandem.

Can you share the orchestrator function code you used to accomplish this for your in-proc solution? Two things I'd like to understand: 
1. Whether there's a feature regression here between the in-process model and the .NET Isolated worker model.
2. Whether there's a simpler model that we could introduce that matches your use case (it's easier for me to see the code than to read a text description).

Thanks in advance.

**Comment by tomseida** [comment]
My pattern was to define a retry policy(ies) in the function's appsettings JSON file. In the example below I have created one retry policy called "Default":

```
    ...
    "ActivityFunctionRetry": {
      "Default": {
        "BackoffCoefficient": 2.0,
        "FirstRetryInterval": "00:00:10",
        "MaxNumberOfAttempts": 10,
        "MaxRetryInterval": "00:05:00"
      }
    },
    ...
```

The retry policy options are added to the service collection using `Configure<ActivityFunctionRetryOptions>()` and are accessed via `IOptionsSnapshot<ActivityFunctionRetryOptions>`. The `ActivityFunctionRetryOptions` class is presented below:

```
public class ActivityFunctionRetryOptions
{
    public double BackoffCoefficient { get; set; } = 1;
    public TimeSpan FirstRetryInterval { get; set; } = TimeSpan.Zero;
    public int MaxNumberOfAttempts { get; set; } = 1;
    public TimeSpan MaxRetryInterval { get; set; } = TimeSpan.MaxValue;
    public TimeSpan RetryTimeout { get; set; } = TimeSpan.MaxValue;
}
```

Below is an orchestration function that receives the `IOptionsSnapshot<ActivityFunctionRetryOptions>` and retrieves the instance named "Default" (name that appears in appsettings JSON file above). When the `RunOrchestrator` method is called a `RetryPolicy` is created via extension method `GetRetryOptions()` (code is presented further below). The `RetryPolicy` has both the timing\attempts of retries and a delegate that evaluates the exception type to determine if a retry should be performed. 

```
public class OF_ReceiveCCCServiceARDetailReport
{
    internal const string FunctionName = nameof(OF_ReceiveCCCServiceARDetailReport);

    private readonly ActivityFunctionRetryOptions _activityFunctionRetryOptions;

    public OF_ReceiveCCCServiceARDetailReport(
        IOptionsSnapshot<ActivityFunctionRetryOptions> activityFunctionRetryOptions)
    {
        _activityFunctionRetryOptions = activityFunctionRetryOptions.Get("Default");
    }

    [FunctionName(FunctionName)]
    public async Task RunOrchestrator(
        [OrchestrationTrigger] IDurableOrchestrationContext context,
        ILogger log)
    {
        OF_ReceiveCCCServiceARDetailReportCommand command = context.GetInput<OF_ReceiveCCCServiceARDetailReportCommand>();
        RetryOptions retryOptions = _activityFunctionRetryOptions.GetRetryOptions((ex) =>
            {
                var innerEx = ex.InnerException ?? ex;

                if (innerEx is NotFoundException) return false;
                if (innerEx is ProcessBatchException) return false;

                return true;
            });

        try
        {
            // ---------------------------------------------------------------------
            // Receive Document
            // ---------------------------------------------------------------------
            AF_ReceiveCCCServiceARDetailReportResponse receiveCCCServiceARDetailReportResponse = 
                await context.CallActivityWithRetryAsync<AF_ReceiveCCCServiceARDetailReportResponse>(
                    AF_ReceiveCCCServiceARDetailReport.FunctionName,
                    retryOptions,
                    new AF_ReceiveCCCServiceARDetailReportCommand(command));
      ...
```

The `GetRetryOptions()` extension method:

```
    public static RetryOptions GetRetryOptions(this ActivityFunctionRetryOptions options, Func<Exception, bool> handle)
    {
        return new(options.FirstRetryInterval, options.MaxNumberOfAttempts)
        {
            BackoffCoefficient = options.BackoffCoefficient,
            MaxRetryInterval = options.MaxRetryInterval,
            RetryTimeout = options.RetryTimeout,
            Handle = handle
        };
    }
```

I think the current retry models suffice. When I wrote the code in pull request #281, I played it safe (i.e., did not want to alter execution for those that are not setting both `RetryPolicy? Policy` and `AsyncRetryHandler? Handler`). I think the more proper, less code version, would have **all** activity functions call `InvokeWithCustomRetryHandler<T>()`. If the caller does not set `RetryPolicy? Policy` then a default policy of _one activity function call_ will be set (i.e., no retries). If the caller does not set `AsyncRetryHandler? Handler`, then no exceptions will be ignored. 

Hope this clarifies how I was using RetryPolicy for "in-process" mode.

**Comment by cgillum** [comment]
@tomseida thanks for this context - I forgot that the previous `RetryOptions` in the in-proc model had a convenient exception type filter feature. It seems like we should add this in as part of the static `RetryPolicy` to provide better parity with the in-prc model. I'll take a look at your PR as well - thanks for linking to that.

**Comment by sacrajit** [comment]
I tried this workaround to unblock till SDK supports it natively

internal static AsyncRetryHandler BuildRetryHandler(CustomRetryOptions options)
{
    var firstRetryInterval = TimeSpan.FromSeconds(options.DefaultActivityRetryOptions.FirstRetryIntervalInSeconds);
    var backoffCoefficient = options.DefaultActivityRetryOptions.BackoffCoefficient;
    var maxRetryInterval = TimeSpan.FromSeconds(options.DefaultActivityRetryOptions.RetryTimeoutInSeconds);
 
    return async (RetryContext rc) =>
    {
        bool shouldRetry = rc.LastAttemptNumber < maxRetryCount;
        bool isRetriable = rc.LastFailure.IsCausedBy<SomeRetryableException>();
 
        if (isRetriable && shouldRetry)
        {
            TimeSpan backOffTime = rc.LastAttemptNumber == 0 ?
                firstRetryInterval :
                TimeSpan.FromSeconds(firstRetryInterval.TotalSeconds * Math.Pow(backoffCoefficient, rc.LastAttemptNumber));
 
            await Task.Delay(backOffTime);
 
            return true;
        }
 
        return false;
    };
}

**Comment by tomseida** [comment]
Created new PR #314 per suggestion of @cgillum.

**Comment by tomseida** [comment]
Closing this issue since PR #314 resolves issue.

**Comment by MarioMajcicaAtABNAMRO** [comment]
I raised the following issue, now I do see this thread. Trying to link it https://github.com/Azure/azure-functions-durable-extension/issues/2908
I'm also a bit confused, is this the project that implements the durable functions for isolated process function model, or is it https://github.com/Azure/azure-functions-durable-extension?

**Comment by LockTar** [comment]
@tomseida If I understand correctly. With your sample when the activity throws `NotFoundException` the function should **not** be retried. So the retry policy should be something like:

```C#
var taskOptions = TaskOptions.FromRetryPolicy(
    new RetryPolicy(
    maxNumberOfAttempts: 3,
    firstRetryInterval: TimeSpan.FromSeconds(5)
    )
    {
        HandleFailure = (TaskFailureDetails taskFailureDetails) =>
        {
            if (taskFailureDetails.ErrorType.Equals("NotFoundException"))
            {
                // Do not retry if the entity was not found
               logger.LogWarning("Do not retry because the error was of the type 'NotFoundException'.");
                return false;
            }
            else
            {
                // Log the exception and retry
                logger.LogError(taskFailureDetails.ErrorMessage);
                return true;
            }
        }
    });
```

**Comment by tomseida** [comment]
@LockTar That is correct, but I would recommend using `IsCausedBy<>()` generic method in place of a string comparison. For example:

```
var taskOptions = TaskOptions.FromRetryPolicy(
    new RetryPolicy(
    maxNumberOfAttempts: 3,
    firstRetryInterval: TimeSpan.FromSeconds(5)
    )
    {
        HandleFailure = (TaskFailureDetails taskFailureDetails) =>
        {
            if (taskFailureDetails.IsCausedBy<NotFoundException>())
            {
                // Do not retry if the entity was not found
                logger.LogWarning("Do not retry because the error was of the type 'NotFoundException'.");
                return false;
            }
            else
            {
                // Log the error message and retry
                logger.LogError("The error message was: {ErrorMessage}. Will retry.", taskFailureDetails.ErrorMessage);
                return true;
            }
        }
    });
```

**Comment by LockTar** [comment]
Thanks! I will look at `IsCausedBy<>()`. 
I also gave some feedback to the docs because this `HandleFailure` is not mentioned there.
